client/data/media/use-delete-media-mutation.js

```diff
@@ -0,0 +1,24 @@
+import { useCallback } from 'react';
+import { useMutation } from 'react-query';
+import wp from 'calypso/lib/wp';
+
+export const useDeleteMediaMutation = ( siteId, queryOptions = {} ) => {
+	const mutation = useMutation(
+		( { mediaId } ) => wp.req.post( `/sites/${ siteId }/media/${ mediaId }/delete` ),
+		{ ...queryOptions }
+	);
+
+	const { mutate } = mutation;
+
+	const deleteMedia = useCallback(
+		( items ) => {
+			const mediaItems = Array.isArray( items ) ? items : [ items ];
```

Comment:
I would be nice if `items` was always an array and if all consumers passed an array. That makes the hook easier to type for example. I think this is a good opportunity to clean this up.

Reply:
One more thing I would change about the API is to make `items` an array of IDs instead of an array of `item`s. And let's make the caller responsible for passing valid IDs, so that we don't need to filter out undefined ones. Then the impl becomes just:
```js
( siteId, mediaIds ) => mediaIds.forEach( ( mediaId ) => mutate( { siteId, mediaId } ) )
```

Reply:
One more idea how to implement deleting of multiple items: move the parallel requests to the mutation function:
```js
const mutation = useMutation( ( { siteId, mediaIds } ) => {
  return Promise.all( mediaIds.map( ( mediaId ) => wpcom.req.post( ... ) ) );
} );
const deleteMedia = ( siteId, mediaIds ) => mutate( { siteId, mediaIds } );
```
This way we avoid calling `mutate` multiple times in parallel, and fields like `mutation.isSuccess` or `.isLoading` have a well-defined meaning: it's not just the last request, but all of them combined.

Reply:
Just for reference I added https://github.com/Automattic/wp-calypso/pull/58573/commits/6ee64fd367d4d7b2cc175dda12f3031f20493e07 which contains the simplifications from your previous comment.

> One more idea how to implement deleting of multiple items […]

A thing which comes to mind when looking at the code is the way those requests are handled in case one or more of them fail and others are successful. Would that still work as is now? I'd consider those independent requests which trigger their own error notices. I can see the problem with `mutation.isSuccess` which does exist now (although it's not used). Maybe another solution is to move the loop to component space and only accept a single `mediaId`? Though, that as well doesn't add a lot of meaning to `mutation.isSuccess` I suppose.

Reply:
If you wanted to be perfect, you'd collect all requests' results with `Promise.allSettled` and:
- for these that succeeded, dispatch the `deleteMediaAction` and `gutenframeUpdateImageBlocks` actions
- if at least one deletion succeeded, dispatch `requestMediaStorage` once to update the storage info. The current code, both before and after this PR, requests the storage for each item, which is a bit inefficient. The actual REST request is deduplicated though, so we don't really issue X requests.
- if any deletion failed, display a notice. The current notice reads "We were unable to delete this media item," and showing it multiple times doesn't make sense, so we could just display one "some items could not be deleted please don't be mad" for all 🙂

Reply:
@jsnajdr I pushed an attempt here: https://github.com/Automattic/wp-calypso/pull/58573/commits/c2c08ed92512beb8d95a40bcd2086a8aa830bf36. It ticks all the boxes from your last comment. Additonally it moves consolidation of multiple requests into one out of the query hook and makes it a responsibility of the consumer. Let me know if you think that's feasible.